client/beefy/src/worker.rs

```diff
@@ -471,15 +470,21 @@ where
 			true,
 		);
 
-		self.gossip_engine.lock().gossip_message(topic::<B>(), encoded_message, false);
+		self.gossip_engine.gossip_message(topic::<B>(), encoded_message, false);
 	}
 
 	/// Wait for BEEFY runtime pallet to be available.
 	async fn wait_for_runtime_pallet(&mut self) {
-		self.client
-			.finality_notification_stream()
-			.take_while(|notif| {
-				let at = BlockId::hash(notif.header.hash());
+		let gossip_engine = &mut self.gossip_engine;
+		let mut finality_stream = self.client.finality_notification_stream();
+		loop {
+			futures::select! {
```

Comment:
Not an expert about the internals of beefy protocol, etc.

But maybe here we have to give priority to the finality_stream? I.e. poll the `gossip_engine` only as a fallback.

Thus use `select_biased`?

Reply:
There's another loop down in `run` that does the same thing (polls the finality stream and the gossip engine) and that also uses `select`, so I just did what it does.

It might make sense to change this, but I don't have the necessary knowledge to be able to make that decision.

Reply:
Yeah. Me neither :-)

But it was just looking for a way to preserve the semantics of the prev implementation, i.e. `wait_for_runtime_pallet` waits for the pallet to be available and if it is available then just return and wait to process the `gossip_engine` messages at a later stage.

Maybe because (if pallet is available) some actions should be taken before start processing the `gossip_engine` messages. For example I see that, in the other loop you are talking about, we wait for the end of mayor syncing [here](https://github.com/paritytech/substrate/blob/b5d36a6bc57c7e0965d0871b194f4e65c26fc15e/client/beefy/src/worker.rs#L535). With this modification we may end-up fetching `gossip_engine` messages before (but maybe there are not drawbacks at all... and this is fine).

Note: the `select_biased` I was talking about may not work as well... because if not active we end up [here](https://github.com/paritytech/substrate/blob/b5d36a6bc57c7e0965d0871b194f4e65c26fc15e/client/beefy/src/worker.rs#L501).

Reply:
it is fine to _process/pump_ `gossip_engine` messages immediately, even while waiting for pallet availability or major syncing, so I wouldn't worry about this